ISOLATE_STDLIB_MODULES.rst

```diff
@@ -0,0 +1,240 @@
+be converted should stay static.
+
+
+Testing
+-------
+
+XXX How should this be tested?
+
+
+Process
+-------
+
+The following process should be added to the Devguide, and stay there until
+all modules are converted.
+Any new findings should be documented there or in the general HOWTO.
+
+Part 1: Preparation
+...................
+
+1. Open a discussion, either on the bug tracker or on Discourse. Involve the
+   module maintainer and/or code owner. Explain the reason and rationale for
+   the changes.
+2. Identify global state performance bottlenecks, if there are such. Create a
+   proof-of-concept implementation and measure the performance impact. `pyperf`
+   is a good tool for benchmarking.
+3. Create an implementation plan. For small modules with few types, a single PR
+   may do the job. For larger modules with lots of types, and possibly also
+   external library callbacks, multiple PR's will be needed.
+
+
+Part 2: Implementation
+......................
+
+Note: this is a suggested implementation plan for a complex module, based on
+lessons learned with other modules. Feel free to simplify it for
+smaller modules.
+
+1. Add Argument Clinic where possible; it enables you to easily use the
+   defining class to fetch module state from type methods.
+2. Prepare for module state; establish a module state `struct`, add an instance
+   as a static global variable, and create helper stubs for fetching the module
+   state.
+3. Add relevant global variables to the module state `struct` and modify code
+   that accesses the global state to use the module state helpers instead. This
+   step may be broken into several PR's.
+4. Where necessary, convert heap types to static types.
+5. Convert the global module state struct to true module state.
+6. Implement multi-phase initialisation.
+
+Preferably, steps 4 through 6 should all land early in a single alpha
+development phase.
+
+
+Backwards Compatibility
+=======================
+
+Extension modules in the standard library will now be loadable more than once.
+For example, deleting such a module from ``sys.modules`` and re-importing it
+will result in a fresh module instance, isolated from any previously loaded
+instances.
+
+This may affect code that expected the previous behavior: globals of
+extension modules were shallowly copied from the first loaded module.
+
+
+Security Implications
+=====================
+
+None known.
+
+
+How to Teach This
+=================
+
+A large part of this proposal is a HOWTO aimed at experienced users,
+which will be moved to the documentation.
+
+Beginners should not be affected.
+
+
+Reference Implementation
+========================
+
+Main branch, particularly commits for these issues:
+
+- https://bugs.python.org/issue40077
+- https://bugs.python.org/issue46417
+- https://bugs.python.org/issue1635741
+
+
+Rejected Ideas
+==============
+
+Keep things as they were and not break anything
```

Comment:
Would not this section make more sense in the subinterpreters or per-interpreter GIL PEPs? It feels kinda out of place here. Also, if we are to keep it, I would rephrase it to something more accurate[^1]. Perhaps _"Do not isolate modules in the stdlib"_.

[^1]: It is not clear to me what "things" is. I would guess that the subject implies _"don't convert static types to heat types"_, _"don't convert global state to module state"_, _"don't apply PEP 630 to the stdlib"_, or _"drop subinterpreter support"_.

Reply:
Yeah, the title should have XXX as well. I guess “Do not isolate modules in the stdlib” works best, because it needs the other changes.

Reply:
We should add a short sentence about why having stdlib extension modules _without_ multi-init support and _with_ global state is a problem; for example, such modules would hamper a lot of Eric's work-in-progress.